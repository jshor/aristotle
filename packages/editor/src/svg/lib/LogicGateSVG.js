import SVGBase from './SVGBase'

const PORT_WIDTH = 30
const STROKE_WIDTH = 2
const WIRE_LENGTH = 40
const LOGIC_GATE_WIDTH = 100
const LOGIC_GATE_HEIGHT = 50

const PATH = {
  OR: 'm 0.13229001,0.13246053 v 0.0165 C 0.71404001,1.1583605 1.05006,2.3240605 1.05006,3.5719605 c 0,1.248 -0.33601999,2.4136 -0.91776999,3.423 v 0.0165 H 3.69589 c 2.5428,0 4.7625,-1.3862 5.9531,-3.4396 -1.1906,-2.0534 -3.4103,-3.43959997 -5.9531,-3.43959997 h -2.6458 z',
  NOR: 'm 9.6570348,3.5685025 c 0,0.2921 -0.237067,0.5291667 -0.529167,0.5291667 -0.2921,0 -0.5291667,-0.2370667 -0.5291667,-0.5291667 0,-0.2921001 0.2370667,-0.5291667 0.5291667,-0.5291667 0.2921,0 0.529167,0.2370666 0.529167,0.5291667 z M 0.12510134,0.12472492 v 0.0165371 C 0.64193176,1.1529323 0.94045405,2.3212541 0.94045405,3.5719608 c 0,1.2508069 -0.29852229,2.4190284 -0.81535271,3.4306986 v 0.016537 H 3.2910264 c 2.2590398,0 4.2310355,-1.3893178 5.2887721,-3.4473361 C 7.5220619,1.5138422 5.5500662,0.12452447 3.2910264,0.12452447 H 0.94048068 Z',
  AND: 'm5.1594 7.0114-5.0271 1e-6v-6.8788l5.0271-3.704e-4c1.8987 0 3.4396 1.5409 3.4396 3.4396 0 1.8987-1.5409 3.4396-3.4396 3.4396z',
  NAND: 'm 9.6770082,3.5685025 c 0,0.2921 -0.237067,0.5291667 -0.529167,0.5291667 -0.2921,0 -0.5291667,-0.2370667 -0.5291667,-0.5291667 0,-0.2921 0.2370667,-0.5291667 0.5291667,-0.5291667 0.2921,0 0.529167,0.2370667 0.529167,0.5291667 z m -4.5176325,3.4429409 -5.02708403,1e-6 V 0.1326548 L 5.1593757,0.1322844 c 1.8986501,0 3.4395832,1.5409332 3.4395832,3.4395833 0,1.8986507 -1.5409331,3.4395837 -3.4395832,3.4395837 z',
  XOR_BASE: 'm0.17336 0.089543c0.54262 1.0146 0.85604 2.186 0.85604 3.4404 0 1.2543-0.31343 2.4258-0.85604 3.4404'
}

const NEGATION_X = {
  NOR: 84,
  XNOR: 84,
  NAND: 78
}

export default class LogicGateSVG extends SVGBase {
  constructor (options) {
    super(options)

    this.inputCount = options.inputCount
    this.gateType = options.gateType
    this.baseLineAttrs = {
      'stroke': this.primaryColor,
      'stroke-linecap': 'square',
      'stroke-width': STROKE_WIDTH,
      'vector-effect': 'non-scaling-stroke'
    }
  }

  /**
   * Renders a set of horizontal wire(s) for the given parameters.
   * 
   * @param {Number} wireCount - number of wires to render
   * @param {Number} x-axis offset
   * @param {Number} svgHeight - total outer height of the SVG
   * @returns {String[]} list of <line> elements
   */
  getWires = (wireCount, x, svgHeight) => {
    const inputHeight = (wireCount - 1) * PORT_WIDTH
    const yOffset = svgHeight / 2 - inputHeight / 2
    const wires = []

    for (let i = 0; i < wireCount; i++) {
      wires.push(this.toSvg('line', {
        x1: x,
        x2: x + WIRE_LENGTH,
        y1: yOffset + i * PORT_WIDTH,
        y2: yOffset + i * PORT_WIDTH,
        ...this.baseLineAttrs
      }))
    }

    return wires
  }

  /**
   * Calculates the locations of the input ports.
   * 
   * @param {Number} svgHeight - total outer height of the SVG
   * @returns {Object<{ x: Number, y: Number, type: String }>[]}
   */
  getInputPorts = (svgHeight) => {
    const inputHeight = (this.inputCount - 1) * PORT_WIDTH
    const yOffset = svgHeight / 2 - inputHeight / 2
    const ports = []

    for (let i = 0; i < this.inputCount; i++) {
      const x = 0
      const y = yOffset + i * PORT_WIDTH

      ports.push({ x, y, type: 'input' })
    }

    return ports
  }

  /**
   * Returns the inner figure (the gate minus the wires or symbols).
   * 
   * @param {Number} x - x-axis offset
   * @param {String} pathData - SVG path of the figure
   * @returns {String} SVG
   */
  getFigureSvg = (x, pathData) => {
    const path = this.toSvg('path', {
      d: pathData,
      fill: this.secondaryColor,
      ...this.baseLineAttrs
    })

    return this.toSvg('svg', {
      height: '100%',
      viewBox: '0 0 16 7.25',
      preserveAspectRatio: 'none',
      x
    }, [path])
  }

  /**
   * Returns the inner figure path for the gate type.
   * 
   * @returns {String} SVG path data
   */
  getPathData = () => {
    switch (this.gateType) {
      case 'AND':
      case 'NAND':
        return PATH.AND
      case 'OR':
      case 'XOR':
      case 'XNOR':
        return PATH.OR
      default:
        return PATH.OR
    }
  }

  /**
   * Determines whether a negation symbol is required for the gate type.
   * 
   * @returns {Boolean}
   */
  isNegated = () => {
    return ['XNOR', 'NOR', 'NAND'].indexOf(this.gateType) !== -1
  }

  /**
   * Renders a negation symbol.
   * 
   * @param {Number} height - total SVG height
   * @returns {String} SVG <circle> element
   */
  getNegation = (height) => {
    const x = NEGATION_X[this.gateType]
    const r = 4

    if (x) {
      return this.toSvg('circle', {
        cx: x,
        cy: height / 2,
        r,
        fill: this.secondaryColor,
        ...this.baseLineAttrs
      })
    }
    return ''
  }

  /**
   * Returns the path, ports and dimensions of the rendered logic gate SVG.
   * 
   * @returns {Object<{ ports: Object[], path: Buffer, width: Number, height: Number }>}
   */
  getSvgData = () => {
    const height = this.inputCount > 2
      ? this.inputCount * PORT_WIDTH + this.inputCount * STROKE_WIDTH
      : LOGIC_GATE_HEIGHT
    const width = LOGIC_GATE_WIDTH
    const pathData = this.getPathData()
    const children = [
      ...this.getWires(this.inputCount, 0, height),
      ...this.getWires(1, width - WIRE_LENGTH, height),
      this.getFigureSvg(20, pathData)
    ]

    if (this.gateType === 'XOR' || this.gateType === 'XNOR') {
      // add the curve at the bottom of XOR gates
      children.push(getFigureSvg(10, PATH.XOR_BASE, 'none'))
    }

    children.push(this.getNegation(height))

    const svg = this.toSvg('svg', {
      width,
      height,
      xmlns: 'http://www.w3.org/2000/svg'
    }, children)

    const path = this.toDataUrl(svg)

    const ports = [
      // input ports
      ...this.getInputPorts(height),
      // single output port
      { x: width, y: height / 2, type: 'output' }
    ]

    return {
      path,
      width,
      height,
      ports
    }
  }
}
