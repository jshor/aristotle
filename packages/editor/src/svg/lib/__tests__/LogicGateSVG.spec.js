import LogicGateSVG from '../LogicGateSVG'

describe('Logic Gate SVG library', () => {
  let renderer

  beforeEach(() => {
    renderer = new LogicGateSVG({
      inputCount: 2,
      gateType: 'NOR'
    })
  })

  describe('getWires()', () => {
    it('should return n line elements', () => {
      const x = 10
      const height = 50
      const wireCount = 3
      const lines = renderer.getWires(wireCount, x, height)

      expect(lines).toHaveLength(wireCount)
      lines.forEach((line) => expect(line).toMatch(/^<line([^>]+)>$/))
    })
  })

  describe('isNegated()', () => {
    it('should return true for NOR', () => {
      renderer.gateType = 'NOR'
      expect(renderer.isNegated()).toEqual(true)
    })

    it('should return true for NAND', () => {
      renderer.gateType = 'NAND'
      expect(renderer.isNegated()).toEqual(true)
    })

    it('should return true for XNOR', () => {
      renderer.gateType = 'XNOR'
      expect(renderer.isNegated()).toEqual(true)
    })

    it('should return false for OR', () => {
      renderer.gateType = 'OR'
      expect(renderer.isNegated()).toEqual(false)
    })
  })

  describe('setInputCount()', () => {
    it('should set `inputCount` to the given value', () => {
      const inputCount = 3

      renderer.inputCount = 2
      renderer.setInputCount(inputCount)

      expect(renderer.inputCount).toEqual(inputCount)
    })
  })

  describe('render()', () => {
    it('should return an array of port definitions', () => {
      const data = renderer.getSvgData()
      expect(data).toHaveProperty('ports')
      expect(Array.isArray(data.ports)).toEqual(true)
    })

    describe('path creation', () => {
      beforeEach(() => {
        jest
          .spyOn(renderer, 'toDataUrl')
          .mockImplementation(s => s)
      })

      function getSvg (gateType) {
        renderer.gateType = gateType
        return renderer.getSvgData().path
      }

      it('should render an OR gate', () => {
        expect(getSvg('OR')).toMatchSnapshot()
      })

      it('should render an NOR gate', () => {
        expect(getSvg('NOR')).toMatchSnapshot()
      })

      it('should render an XOR gate', () => {
        expect(getSvg('XOR')).toMatchSnapshot()
      })

      it('should render an XNOR gate', () => {
        expect(getSvg('XNOR')).toMatchSnapshot()
      })

      it('should render an AND gate', () => {
        expect(getSvg('AND')).toMatchSnapshot()
      })

      it('should render an NAND gate', () => {
        expect(getSvg('NAND')).toMatchSnapshot()
      })
    })
  })
})
