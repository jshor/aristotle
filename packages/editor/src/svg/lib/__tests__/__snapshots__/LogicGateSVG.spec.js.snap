// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic Gate SVG library render() path creation should render an AND gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m5.1594 7.0114-5.0271 1e-6v-6.8788l5.0271-3.704e-4c1.8987 0 3.4396 1.5409 3.4396 3.4396 0 1.8987-1.5409 3.4396-3.4396 3.4396z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg></svg>"`;

exports[`Logic Gate SVG library render() path creation should render an NAND gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m5.1594 7.0114-5.0271 1e-6v-6.8788l5.0271-3.704e-4c1.8987 0 3.4396 1.5409 3.4396 3.4396 0 1.8987-1.5409 3.4396-3.4396 3.4396z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg><circle cx=\\"78\\" cy=\\"25\\" r=\\"4\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg>"`;

exports[`Logic Gate SVG library render() path creation should render an NOR gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m 0.13229001,0.13246053 v 0.0165 C 0.71404001,1.1583605 1.05006,2.3240605 1.05006,3.5719605 c 0,1.248 -0.33601999,2.4136 -0.91776999,3.423 v 0.0165 H 3.69589 c 2.5428,0 4.7625,-1.3862 5.9531,-3.4396 -1.1906,-2.0534 -3.4103,-3.43959997 -5.9531,-3.43959997 h -2.6458 z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg><circle cx=\\"84\\" cy=\\"25\\" r=\\"4\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg>"`;

exports[`Logic Gate SVG library render() path creation should render an OR gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m 0.13229001,0.13246053 v 0.0165 C 0.71404001,1.1583605 1.05006,2.3240605 1.05006,3.5719605 c 0,1.248 -0.33601999,2.4136 -0.91776999,3.423 v 0.0165 H 3.69589 c 2.5428,0 4.7625,-1.3862 5.9531,-3.4396 -1.1906,-2.0534 -3.4103,-3.43959997 -5.9531,-3.43959997 h -2.6458 z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg></svg>"`;

exports[`Logic Gate SVG library render() path creation should render an XNOR gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m 0.13229001,0.13246053 v 0.0165 C 0.71404001,1.1583605 1.05006,2.3240605 1.05006,3.5719605 c 0,1.248 -0.33601999,2.4136 -0.91776999,3.423 v 0.0165 H 3.69589 c 2.5428,0 4.7625,-1.3862 5.9531,-3.4396 -1.1906,-2.0534 -3.4103,-3.43959997 -5.9531,-3.43959997 h -2.6458 z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"10\\"><path d=\\"m0.17336 0.089543c0.54262 1.0146 0.85604 2.186 0.85604 3.4404 0 1.2543-0.31343 2.4258-0.85604 3.4404\\" fill=\\"none\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg><circle cx=\\"84\\" cy=\\"25\\" r=\\"4\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg>"`;

exports[`Logic Gate SVG library render() path creation should render an XOR gate 1`] = `"<svg width=\\"100\\" height=\\"50\\" xmlns=\\"http://www.w3.org/2000/svg\\"><line x1=\\"0\\" x2=\\"40\\" y1=\\"10\\" y2=\\"10\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"0\\" x2=\\"40\\" y1=\\"40\\" y2=\\"40\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><line x1=\\"60\\" x2=\\"100\\" y1=\\"25\\" y2=\\"25\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"20\\"><path d=\\"m 0.13229001,0.13246053 v 0.0165 C 0.71404001,1.1583605 1.05006,2.3240605 1.05006,3.5719605 c 0,1.248 -0.33601999,2.4136 -0.91776999,3.423 v 0.0165 H 3.69589 c 2.5428,0 4.7625,-1.3862 5.9531,-3.4396 -1.1906,-2.0534 -3.4103,-3.43959997 -5.9531,-3.43959997 h -2.6458 z\\" fill=\\"undefined\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg><svg height=\\"100%\\" viewBox=\\"0 0 16 7.25\\" preserveAspectRatio=\\"none\\" x=\\"10\\"><path d=\\"m0.17336 0.089543c0.54262 1.0146 0.85604 2.186 0.85604 3.4404 0 1.2543-0.31343 2.4258-0.85604 3.4404\\" fill=\\"none\\" stroke=\\"undefined\\" stroke-linecap=\\"square\\" stroke-width=\\"2\\" vector-effect=\\"non-scaling-stroke\\" /></svg></svg>"`;
